KEP-4322: Add cluster inventory definition (#4620)
kubernetes/enhancements: keps/sig-multicluster/4322-cluster-inventory/README.md

Currently, there is a lack of a standardized approach to define a
cluster inventory. However, with the growing number of users managing
multiple clusters and deploying applications across them, projects like
[Open Cluster Management (OCM)](https://open-cluster-management.io/),

> [Clusternet](https://clusternet.io/) or [Kubernetes Fleet Manager](https://github.com/Azure/fleet)

Reviewer: Hi @ryanzhang-oss ,
It doesn't make sense to move Karmada out of the list.

Reviewer 2: I think it needs to be added, we even mention karmada at the kubecon talk about this API

  clusters under management.

### Terminology

> - **Cluster Inventory**: A collection of clusters governed by a single authority. These clusters SHOULD

Reviewer 3: 1. Should we clarify that the ClusterInventory is the List of ClusterProfiles in the same namespace and that ClusterProfile is namespaced for that reason? 
2. I'd say "MUST adhere to sameness"; because it could break controller assumptions.
3. I'd skip the definition of sameness in favor of just the link (because I disagree with "generally behave the same", it's a "same identity"... they are the same thing as far as other services care)
4. You don't mention ClusterSet. It sounds like we could replace sameness by clusterset since its already defined? Or are we giving up on ClusterSet? And instead re-assigning sameness to be at the "Namespace of ClusterProfiles"?

  clusters under management.

### Terminology
- **Cluster Manager**: An entity that creates the inventory cluster API 
  object per member cluster, and keeps their status up-to-date. Each
  cluster manager SHOULD be identified with a unique name. Each cluster
  inventory resource SHOULD be managed by only one cluster manager. Examples
  of cluster manager are projects like [OCM](https://open-cluster-management.io/),
  [Clusternet](https://clusternet.io/) or [Kubernetes Fleet Manager](https://github.com/Azure/fleet)

> - **Cluster Inventory**: A collection of clusters governed by a single cluster manager.

Reviewer 4: Should we need to give a quick mention to ClusterSet? https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#terminology

  clusters under management.

### Terminology
- **Cluster Inventory**: A conceptual term referring to a collection of clusters.

- **Member Cluster**: A kubernetes cluster that is part of a cluster inventory.

- **Cluster Manager**: An entity that creates the ClusterProfile API object per member cluster,
  and keeps their status up-to-date. Each cluster manager MUST be identified with a unique name.  
  Each ClusterProfile resource SHOULD be managed by only one cluster manager. A cluster manager SHOULD 
  have sufficient permission to access the member cluster to fetch the information so it can update the status

>   of the ClusterProfile API resource. Examples of cluster manager are projects like [OCM](https://open-cluster-management.io/),

Reviewer 5: Are these examples of a "Cluster Manager" which "creates ClusterProfile API objects"? I don't think any of them use "ClusterProfile API" yet.

Author: good point, we haven't implemented those yet (since the clusterProfile API is still not well defined)

>   [Clusternet](https://clusternet.io/) or [Kubernetes Fleet Manager](https://github.com/Azure/fleet)

Reviewer: ditto
Karamda should be here.  Here is the website: https://karmada.io/

Reviewer 2: agree

further enhance the usability and adoption of different cluster
manager.

### Notes/Constraints/Caveats
#### What's the relationship between the ClusterProfile API and Cluster Inventory?
The ClusterProfile API represents a single member cluster in a cluster inventory.

#### What's the relationship between a cluster inventory and clusterSet?
A cluster inventory may or may not represent a ClusterSet. A cluster inventory is considered a [clusterSet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#terminology)
if all its member clusters adhere to the [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md) principle.

> In other terms, a clusterSet is a special type of cluster inventory while the reverse is not true.

Reviewer 6: I don't think this line ins necessary and I find it a bit confusing. A special type of cluster inventory may represent a cluster set, but cluster set as a concept exists separately from cluster inventory, and is not a special case of CI

further enhance the usability and adoption of different cluster
manager.

### Notes/Constraints/Caveats

#### How to Utilize the API
This section provides guidance on the recommended way to interact with the API. It is suggested that
all clusterProfile objects, which represent a single cluster in the inventory, should be emitted

> on a dedicated management/hub Kubernetes cluster. This approach allows consumers to have a single

Reviewer 3: This approach allows "Multicluster controllers to have a global view of the clusters of the ClusterSet".

Author: I would like to not assume that the consumer is a controller, it could be some UI code, script or even human.

integration point to access all the information within a cluster inventory. A controller can be run 
on the management or hub cluster to offer high-level functionalities over this inventory of clusters.

> #### Approaches for Structuring Clusters within an Inventory

Reviewer 3: I'm not sure I understand how this section is worded. The namespace is to make explicit groups of ClusterProfiles and allow multiple groups.

It is a requirement at this point I think: "Allow multiple disjoints groups of clusters".
And the only way to support that in K8s is via a Namespace.

I think it's not a recommendation, it's a mandate.

Author: This is within an inventory.  The recommended way is to place them in the same namespace. Do you want to mandate that?

further enhance the usability and adoption of different cluster
manager.

### Notes/Constraints/Caveats
#### What's the relationship between the clusterProfile API and cluster Inventory?
The clusterProfile API represents a single member cluster in a cluster Inventory.

#### What's the relationship between a cluster Inventory and clusterSet?
A cluster inventory is considered a clusterSet if all its member clusters adhere to the 
[namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md) principle.

> 

Reviewer 4: Do we need to add:

```
#### What's the relationship between a ClusterProfile API and ClusterSet?
```

Author: I thought it was transitive as we defined the relationship between ClusterProfile API <-> cluster Inventory and cluster Inventory <-> clusterSet but I think we can be more clear.

Reviewer 3: That still works here I think, given the definition of ClusterInventory to be a collection of ClusterProfile.
A clusterInventory could be taken at the hub-cluster level and still each namespace to be a clusterSet.

I think the ClusterProfile is within a Namespace, then one can consume the inventory at the namespace or hub-cluster level (sum of all namespaces).
For cluster-set, we're implying the namespace level is the way and adding an extra requirement (all clusters in a namespace-level inventory are part of the same clusterset)

further enhance the usability and adoption of different cluster
manager.

### Notes/Constraints/Caveats
#### What's the relationship between the ClusterProfile API and Cluster Inventory?
The ClusterProfile API represents a single member cluster in a cluster inventory.

#### What's the relationship between a cluster inventory and clusterSet?
A cluster inventory may or may not represent a ClusterSet. A cluster inventory is considered a [clusterSet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#terminology)

> if all its member clusters adhere to the [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md) principle.

Reviewer 7: another restriction is a cluster can only be in one clusterset. There is not such restriction for inventory

> In other terms, a clusterSet is a, special type of, cluster inventory while the reverse is not true.

Reviewer 8: extra commas here?

```
s|a, special type of, cluster inventory|a special type of cluster inventory
```

#### How should the API be consumed?
We recommend that all ClusterProfile objects within the same cluster inventory reside on 
a dedicated hub Kubernetes cluster. This approach allows consumers to have a single integration 
point to access all the information within a cluster inventory. Additionally, a multi-cluster aware
controller can be run on the hub cluster to offer high-level functionalities over this inventory of clusters.

####  How should we organize ClusterProfile objects on a hub cluster?

> While there are no strict requirements, we suggest making the ClusterProfile API a namespace-scoped object. 

Reviewer 2: we strongly recommend

> This approach allows users to leverage Kubernetes' native namespace-based RBAC if they wish to restrict access to 

Author: I wonder how one can use namespace based RBAC to control access of certain cluster profiles if all the ClusterProfiles in a clusterInventory are in the same namespace?

certain clusters within the inventory.  

However, if the cluster inventory is a clusterSet, all its ClusterProfile objects MUST be placed in the namespace 
bare the name of the clusterSet. In addition, the namespace must have a label with the key "clusterset.multicluster.x-k8s.io"
and the value as the name of the clusterSet. It's important to note that this means users would lose the ability to apply Kubernetes' 
native namespace-based RBAC within a clusterSet since Kubernetes does not support nested namespaces.

#### Uniqueness of the ClusterProfile object
While there are no strict requirements, we recommend that there is only one ClusterProfile object representing any member cluster
on a hub cluster. 

> However, a ClusterProfile object can only be in one ClusterSet since the namespace sameness property is transitive, therefore 

Reviewer 2: ```suggestion
However, a ClusterProfile object should only be in one ClusterInventory since the namespace sameness property is transitive, therefore 
```

Author: Cluster inventory doesn't have namespace sameness property.

further enhance the usability and adoption of different cluster
manager.

### Notes/Constraints/Caveats
#### What's the relationship between the ClusterProfile API and Cluster Inventory?
The ClusterProfile API represents a single member cluster in a cluster inventory.

#### What's the relationship between a cluster inventory and clusterSet?
A cluster inventory may or may not represent a ClusterSet. A cluster inventory is considered a [clusterSet](https://github.com/kubernetes/enhancements/tree/master/keps/sig-multicluster/1645-multi-cluster-services-api#terminology)
if all its member clusters adhere to the [namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md) principle.
In other terms, a clusterSet is a special type of cluster inventory while the reverse is not true.
Note that a cluster can only be in one ClusterSet while there is not such restriction for a cluster inventory.

#### How should the API be consumed?
We recommend that all ClusterProfile objects within the same cluster inventory reside on 

> a dedicated hub Kubernetes cluster. This approach allows consumers to have a single integration 

Reviewer 5: Have we defined "hub Kubernetes cluster" somewhere?

Reviewer: Good question. Not yet, maybe we don't need to define what is the hub Kubernetes cluster.

Given the ClusterProfile is created by the `cluster manager`, so I guess here we can say the cluster manager organizes the ClusterProfile.

point to access all the information within a cluster inventory. Additionally, a multi-cluster aware
controller can be run on the hub cluster to offer high-level functionalities over this inventory of clusters.

####  How should we organize ClusterProfile objects on a hub cluster?
While there are no strict requirements, we recommend making the ClusterProfile API a namespace-scoped object. 
This approach allows users to leverage Kubernetes' native namespace-based RBAC if they wish to restrict access to 
certain clusters within the inventory.  

However, if a cluster inventory represents a ClusterSet, all its ClusterProfile objects MUST be part of the same clusterSet
and namespace must be used as the grouping mechanism. In addition, the namespace must have a label with the key "clusterset.multicluster.x-k8s.io"
and the value as the name of the clusterSet. It's important to note that this means users would lose the ability to apply Kubernetes' 

> native namespace-based RBAC within a clusterSet since Kubernetes does not support nested namespaces.

Reviewer 5: This sentence is confusing. ClusterSet and ClusterProfiles are just objects in a namespace. Native RBAC can be applied to the objects. I think this sentence is trying to say that native namespace RBAC in the hub cluster cannot be applied to the clusters that these objects refer to. Is that right?

The clusterProfile API represents a single member cluster in a cluster Inventory.

#### What's the relationship between a cluster Inventory and clusterSet?
A cluster inventory is considered a clusterSet if all its member clusters adhere to the 
[namespace sameness](https://github.com/kubernetes/community/blob/master/sig-multicluster/namespace-sameness-position-statement.md) principle.

#### How should the API be consumed?
We recommend that all clusterProfile objects within the same cluster inventory reside on 
a dedicated hub Kubernetes cluster. This approach allows consumers to have a single integration 
point to access all the information within a cluster inventory. Additionally, a multi-cluster aware
controller can be run on the hub cluster to offer high-level functionalities over this inventory of clusters.

####  How should we organize clusterProfile objects on a hub cluster?
While there are no strict requirements, we suggest making the clusterProfile API a namespace-scoped object. 
This approach allows users to leverage Kubernetes' native namespace-based RBAC if they wish to restrict access to 
certain clusters within the inventory.  

> However, if the cluster inventory is a clusterSet, all the clusterProfile objects MUST be placed in the same namespace. 

Reviewer 3: The name of the namespace doesn't really matter if you have the label.

I think what's important with clusterset is guaranteeing that all clusters can be considered "having sameness" for a user, because that enables a lot of management simplicity. The requirement is that all clusters are part of the same cluster, not necessarily that they are all there.

use cases I can think of:
- regional management of clusterset; if one wanted localized management of clusters. Example: you have 3 clusters in europe and 3 clusters in us. They are part of the same clusterset but you want a regional hub-cluster. You would have 2 hub cluster that each manage 1 inventory of 3 clusters. The two regional clusters would have to be guaranteed by something global that they wont break sameness (maybe a global hub-cluster?).
- isolation of management: let's say we've given a inventory consumer the ability to manage N namespaces within some clusterProfiles, we could have an inventory of clusters thats a subset of the ClusterSet 
- migration: if we needed to migrate a clusterset, we will need to move one cluster at a time to the new representation of the clusterset (but without breaking the ClusterSet rules)
- timing of ON/OFF: clusters may come and go and join/leave the set as they are managed, so the completeness of the set may vary

Obviously a consumer of the API would be limited by the set of profiles they see and it may impact behaviors if the clusterset is not complete, but I think the main rule that we care about here is that the consumer can treat the clusters as being part of the same clusterset.

In addition, the namespace must have a label with the key "clusterset.multicluster.x-k8s.io" and the value as the name 
of the clusterSet. It's important to note that this means users would lose the ability to apply Kubernetes' 
native namespace-based RBAC within a clusterSet since Kubernetes does not support nested namespaces.

#### Uniqueness of the clusterProfile object
While there are no strict requirements, we recommend that there is only one clusterProfile object representing a cluster
on a hub cluster. 

> However, the clusterProfile objects in two namespaces that contain two clusterSets must be disjoint

Author: I think I am explaining the reason behind the rule

further enhance the usability and adoption of different cluster
manager.

### Notes/Constraints/Caveats

#### How to Utilize the API
This section provides guidance on the recommended way to interact with the API. It is suggested that
all clusterProfile objects, which represent a single cluster in the inventory, should be emitted
on a dedicated management/hub Kubernetes cluster. This approach allows consumers to have a single
integration point to access all the information within a cluster inventory. A controller can be run 
on the management or hub cluster to offer high-level functionalities over this inventory of clusters.

#### Approaches for Structuring Clusters within an Inventory
This section provides recommendations on the optimal way to structure clusters within an inventory. 
It's important to note that while there are no strict requirements, we make the clusterProfile API a 
namespace-scoped object. This approach allows users to leverage Kubernetes' native namespace-based
RBAC if they wish to restrict access to certain sub-section of clusters with in the inventory.

#### Strategies for Organizing Multiple Cluster Inventories

> This section offers guidance on the optimal methods for organizing multiple cluster inventories. While there are no
> strict requirements, if users opt to use a single management/hub Kubernetes cluster to manage multiple cluster inventories, 
> it is strongly recommended to place all the clusterProfile objects belonging to the same cluster inventory in the same namespace.
> This approach enforces the principle of namespace sameness within each cluster inventory. However, it's important to note
> that this means users would lose the ability to apply Kubernetes' native namespace-based RBAC within an inventory since
> Kubernetes does not support nested namespaces.

Reviewer 2: ```suggestion
This section aims to guide the best practices for organizing multiple cluster inventories. While there are no strict requirements, it is recommended that users use a single management/hub Kubernetes cluster to manage multiple cluster inventories. If this approach is used, it is strongly advised to keep all the `ClusterProfile` objects of the same cluster inventory in the same namespace. This practice ensures namespace sameness within each cluster inventory. However, it is important to note that this approach will result in the inability to apply Kubernetes' native namespace-based RBAC within an inventory since Kubernetes does not support nested namespaces.
```
